src/integer_mod_q/z_q/arithmetic/pow.rs

```diff
@@ -0,0 +1,127 @@
+// Copyright © 2023 Niklas Siemer
+//
+// This file is part of qFALL-math.
+//
+// qFALL-math is free software: you can redistribute it and/or modify it under
+// the terms of the Mozilla Public License Version 2.0 as published by the
+// Mozilla Foundation. See <https://mozilla.org/en-US/MPL/2.0/>.
+
+//! This module provides an implementation of the [`Pow`] trait for [`Zq`].
+
+use crate::{
+    error::MathError,
+    integer::Z,
+    integer_mod_q::Zq,
+    macros::for_others::{implement_for_others, implement_for_owned},
+    traits::Pow,
+};
+use flint_sys::fmpz_mod::fmpz_mod_pow_fmpz;
+
+impl Pow<&Z> for &Zq {
+    type Output = Zq;
+
+    /// Raises the value of `self` to the power of an integer `exp`.
+    ///
+    /// Parameters:
+    /// - `exp`: specifies the exponent to which the value is raised
+    ///
+    /// Returns the value of `self` powered by `exp` as a new `Output` instance.
+    ///
+    /// # Example
+    /// ```
+    /// use qfall_math::integer::Z;
+    /// use qfall_math::integer_mod_q::Zq;
+    /// use qfall_math::traits::*;
+    ///
+    /// let base = Zq::try_from((2, 9)).unwrap();
+    /// let exp = Z::from(4);
+    ///
+    /// let powered_value = base.pow(&exp).unwrap();
+    ///
+    /// let cmp = Zq::try_from((7, 9)).unwrap();
+    /// assert_eq!(cmp, powered_value);
+    /// ```
+    ///
+    /// # Errors and Failures
+    /// - Returns a [`MathError::InvalidExponent`] if the provided exponent is
+    /// negative and the base value of `self` is not invertible.
+    fn pow(self, exp: &Z) -> Result<Self::Output, MathError> {
+        let mut out = self.clone();
+        if 0 == unsafe {
```

Comment:
Still only takes reference, but should fix the documentation. Only left is to adapt the macro call and trait definition.
```suggestion
impl Pow<&Z> for Zq {
    type Output = Zq;

    /// Raises the value of `self` to the power of an integer `exp`.
    ///
    /// Parameters:
    /// - `exp`: specifies the exponent to which the value is raised
    ///
    /// Returns the value of `self` powered by `exp` as a new `Output` instance.
    ///
    /// # Example
    /// ```
    /// use qfall_math::integer::Z;
    /// use qfall_math::integer_mod_q::Zq;
    /// use qfall_math::traits::*;
    ///
    /// let base = Zq::try_from((2, 9)).unwrap();
    /// let exp = Z::from(4);
    ///
    /// let powered_value = base.pow(&exp).unwrap();
    ///
    /// let cmp = Zq::try_from((7, 9)).unwrap();
    /// assert_eq!(cmp, powered_value);
    /// ```
    ///
    /// # Errors and Failures
    /// - Returns a [`MathError::InvalidExponent`] if the provided exponent is
    /// negative and the base value of `self` is not invertible.
    fn pow(&self, exp: &Z) -> Result<Self::Output, MathError> {
        let mut out = self.clone();
        if 0 == unsafe {
```

Reply:
It's intentional that the base element needs to be borrowed. The syntax of `(&self).pow(exp)` is horrible to write. Hence, I rather do not give the option to pass it in as an owned type and patronize the user a bit for a much simpler way to write it `self.pow(exp)` and not get any problems afterwards. This can probably be changed in the future, if we add a special syntax like `**` or `^` to power a value, but for now I treated it for `gcd`, `xgcd`, and `distance` the same way and think it's the right design decision for now.

Reply:
the syntax of (&self).pow(exp) is not needed you can still write the other option, it is not enforced to do it, hence I still think my suggestion is better, since it does not change syntax allows for link highlighting in the documentation

Reply:
Oh, I just understood what you wanted to tell me. I'll try to figure it out.

Reply:
Should be done now.